### PR TITLE
Module localizations

### PR DESCRIPTION
**Pull request by HechtiDerLachs**
This is work in progress for the implementation of modules over localizations of multivariate polynomial rings. 

The basis is due to @AlexD97 and @jankoboehm  and I try to gradually implement the backends of their generic code in the file `src/Modules/LocalizedModules.jl`.

### CODE REVIEW

File: test/Modules/LocalizedModules.jl [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+  A = L[x 0 1//x; 0 (y-1)//x (y-1)^2//x]
+  B = L[x^2 0 1; 0 (x+1)*(y-1) (x+1)*(y-1)^2]
+  M = SubQuo(F, A, B)
+
+  g = hom(F, M, [M[1], M[2], x*M[2]])
+  Kg, inc_Kg = kernel(g)
+
+  v = x*F[2]-F[3] 
+  groebner_basis(M)
```
Comment (HechtiDerLachs) [inline test/Modules/LocalizedModules.jl:21]: @AlexD97 : For some reason, the caching of the gbs doesn't work. It seems that it doesn't recognize the orderings as keys. Try to execute this file and let me know if you have any solutions for this.

Comment (AlexD97) [reply test/Modules/LocalizedModules.jl:21]: The problem is that equality for module orderings isn’t implemented yet.
CC: @tthsqe12

Comment (tthsqe12) [reply test/Modules/LocalizedModules.jl:21]: Hmm, I am actually not sure how this equality could be implemented.

Comment (thofma) [reply test/Modules/LocalizedModules.jl:21]: Are the mutable? Shouldn't it fallback to `===`? Is this not enough?

**Comment by HechtiDerLachs** [comment]
@jankoboehm @AlexD97 : Please look at the last commit in the file [on localized modules](https://github.com/oscar-system/Oscar.jl/pull/1217/commits/7c8b352d0d4bdd7ad913b7a8be1d739d2e72ff8a#diff-8cabcb925da82b2f0d056800931e235b5189e835a5eff6fd0a05d44d9734efb3). 

This is a prototype for how one could implement a very generic kernel routine for maps of modules over localizations of polynomial rings. As suggested by our previous discussions, this translates the question into the setting of an associated map on free modules over the `base_ring`, computes the intermediate result by means of your high-level routines and then translates back into the localization.

One thing that I found to be necessary is the following. @thofma has extended the maps from polynomial rings `P -> R` so that any target ring `R` is admissible. This provided great flexibility that was really needed in some parts of the code. I think with the modules, we now also have arrived at a point where we need something similar. As for now, there is no type to contain the canonical map of free modules `F -> F[U^{-1}]` from modules to their localization. Likewise for modules over quotient rings. 

To this end, I have made a [draft](https://github.com/oscar-system/Oscar.jl/pull/1217/commits/7c8b352d0d4bdd7ad913b7a8be1d739d2e72ff8a#diff-8cabcb925da82b2f0d056800931e235b5189e835a5eff6fd0a05d44d9734efb3R317) for a type that 

 * allows for the flexibility to include a change of base rings `R -> S` to also contain maps `F -> M` from an `R`-module `F` to an `S`-module `M`;
 * falls back to the previously available functionality for the case `R = S`. 

I do not intend to introduce this as a new type in the long run! Rather, I would like you to have a look at it and discuss how the existing types can be widened accordingly. 

CC: @wdecker

**Comment by HechtiDerLachs** [comment]
In order to have a better overview over the essential requirements for localizations of modules, I started implementing the approach from this paper [here](https://arxiv.org/abs/1709.08180) by Sebastian Posur. It describes a general interface for how to make a localization of a ring computable, provided the original ring was. 

I guess, the methods there could use some fine-tuning and more caching. But it could serve us as a generic fallback and nothing prevents us from overwriting the methods with faster ones for particular cases.

**Comment by HechtiDerLachs** [comment]
We now have a very nicely working generic implementation for localizations of modules which includes caching of groebner basis in the case of localizations of multivariate polynomial rings. 

There are some open questions that I would like to discuss, see the [documentation file](https://github.com/oscar-system/Oscar.jl/pull/1217/files#diff-5792bb999a016339d7747b4aab2b85934fd5a80a6f5a31af1a8aa8f1283a491e)

**Comment by jankoboehm** [comment]
Goes in the right direction. I like that as a very nice generic implementation for situations where we do not have anything better (localization at points, hyperplanes, localization at prime ideals,...). So we should branch in special implementations where applicable. I don't think it is efficient to reduce everything to the localization problem. Another point is the following:

It is of course a triviality from first year math, that you can do anything once you can solve linear systems which means that you need kernel and particular solution. In fact, with some insight in the fundamental algorithms, the particular solution can be reduced again to kernel.

However, represents_element is typically a division with remainder which is much more efficient than coordinates which amounts to a lift. And lift can be done better than using syz. So in our generic interface we want to provide when ever possible all three basic functions:

represents_element
coordinates
kernel

though only the last two or even one will be sufficient.

When constructing a localization we also want to use the respective functions of the underlying polynomial ring.

Btw, our code should also work if the original ring is itself a localization.

Btw, the generic interface must of course accommodate more than just localizations, e.g. also quotient rings.

**Comment by HechtiDerLachs** [comment]
@jankoboehm : What I was looking for is a clear interface that would say "The modules work over `R` whenever the methods `represents_element`, `coordinates` and `kernel` are implemented." Does this exist, yet? If yes, do we have a test suite for this generic module code? Then one could start with tests for implementations of the three methods over arbitrary rings and subsequently their localizations whenever they implement a solution to the localization problem.

I found the generic code to be surprisingly efficient. However, I agree with you that specific methods should be overwritten with faster ones in special cases. But in any case, this generic implementation provides the framework for the programmer, in which such implementations can take place: `base_ring_module`s for free modules, `(pre_)saturated_module`s for the `SubQuo`s and `clear_denominators` for matrices and elements. 

The translation to the 'Computer-algebra-side' is made implicitly, but should be understandable and traceable from the documentation in the source code and Posur's paper. Personally, I do not see any necessity to introduce a new data structure for such translations in general (referring to some of our discussions elsewhere). But if anyone wants to try to introduce such a thing in a reasonable way: Please, take it from here and go for it!

**Comment by jankoboehm** [comment]
You can use the module code through the abstraction layer of providing represents_element, coordinates and kernel. This is what you are doing right? 

In doing so, in your implementation you have to completely provide on your own the computational workhorse for that (Singular communication, caching etc). If you use computations over the underlying  polynomial ring, that makes it easier, of course, since everything is there already. 

The point of a another computational abstraction layer is to assist the user with tools when plugging in their favorite problem into the framework so that they do not have to make their hands dirty with caching and Singular communication. 

It also makes it easier to debug, which is rather difficult if you tell the user do whatever you want. 

It gives you also much more flexibility in how you use Singular then just using computations over polynomial rings. 

And it allows for easy performance analysis. What examples did you try with the generic code and how do they compare in performance with the direct approaches for localization, say at points and hyperplanes?

**Comment by jankoboehm** [comment]
And it of course avoids code duplication (e.g. when implementing localization using the standard approaches, modules over quotient rings etc).

**Comment by HechtiDerLachs** [comment]
I'm not sure whether this has become clear already: The code for localizations here does not need any workhorse for itself! It translates every problem in the localization ``S`` to a problem over the base ring ``R``, assuming that the modules over ``R`` know how to do the required computations. 

So, from what I understand, what you have in mind looks not so much like a particular requirement for localizations, but it looks like another interface for using or translating to `Singular`. One that does the job of `BiPolyArray` or `ModuleGens` in a more general (and probably more coherent) way. 

I see the where the wish for such a data structure comes from: The manual filling of `BiPolyArrays` and the like is a rather tedious task and doing such implementations over and over again for every new specific case (rings, modules, quotient rings,...) is prone to introducing bugs. So it would be nice to have another, universal wrapper for passing from the Oscar to the Singular side. 

But personally, I do not have a sufficiently good overview to realize any such bundle of data structures and methods that is general enough to accommodate all the current and future use cases. Plus: A lot of work has already been put into wrapping Singular rings and modules for Oscar in the first place. Do we really need/want another conversion layer? Or is there, in the end, no other way than to manually get ones hands dirty with the conversions case by case? 

Either way: I am optimistic that localizations can do very well without any direct interface to Singular. At least for the moment.

**Comment by fieker** [comment]
On Wed, Apr 27, 2022 at 06:41:36AM -0700, Matthias Zach wrote:
> I'm not sure whether this has become clear already: The code for localizations here does not need any workhorse for itself! It translates every problem in the localization ``S`` to a problem over the base ring ``R``, assuming that the modules over ``R`` know how to do the required computations. 
> 
> So, from what I understand, what you have in mind looks not so much like a particular requirement for localizations, but it looks like another interface for using or translating to `Singular`. One that does the job of `BiPolyArray` or `ModuleGens` in a more general (and probably more coherent) way. 
> 
> I see the where the wish for such a data structure comes from: The manual filling of `BiPolyArrays` and the like is a rather tedious task and doing such implementations over and over again for every new specific case (rings, modules, quotient rings,...) is prone to introducing bugs. So it would be nice to have another, universal wrapper for passing from the Oscar to the Singular side. 
> 
> But personally, I do not have a sufficiently good overview to realize any such bundle of data structures and methods that is general enough to accommodate all the current and future use cases. Plus: A lot of work has already been put into wrapping Singular rings and modules for Oscar in the first place. Do we really need/want another conversion layer? Or is there, in the end, no other way than to manually get ones hands dirty with the conversions case by case? 
> 
> Either way: I am optimistic that localizations can do very well without any direct interface to Singular. At least for the moment.
> 
Also bear in mind that Christian is working on haveing some
Singular-free Groebner engine(s) in some cases. So any interface that
ties us to Singular is not helpful.

Do we want to start, yet-another-module class?
I can see, generic, a few independent classes of modules which will
share the identical mathematical interface, but not the implementation:
 - modules via a strong rref (over fields and euclidean rings)
 - modules via Groebner
 - other, e.g. modules over Dedekind domains which are genuinely
   different and should not be forced to use Groebner

Of couse, one COULD use the kernel and coordinates only interface to do
everything, but that is not efficient.
in case 1: rref is (usually) so fast that any of the SubQuo ideas are
possible but pointless: I have access to a false unique version, so why
avoid that?

in case 3: of course, I could work with generators, but why waste the
possibility of using the "pseudo generators"? I will be as unique as in
case 1 and almost as fast. Coordinates will be pointless here as I do
not have generators - or possible the coordinates are not in the ring...

What is the goal here?

**Comment by jankoboehm** [comment]
> I'm not sure whether this has become clear already: The code for localizations here does not need any workhorse for itself! It translates every problem in the localization `S` to a problem over the base ring `R`, assuming that the modules over `R` know how to do the required computations.
> 
> So, from what I understand, what you have in mind looks not so much like a particular requirement for localizations, but it looks like another interface for using or translating to `Singular`. One that does the job of `BiPolyArray` or `ModuleGens` in a more general (and probably more coherent) way.
> 
> I see the where the wish for such a data structure comes from: The manual filling of `BiPolyArrays` and the like is a rather tedious task and doing such implementations over and over again for every new specific case (rings, modules, quotient rings,...) is prone to introducing bugs. So it would be nice to have another, universal wrapper for passing from the Oscar to the Singular side.
> 
> But personally, I do not have a sufficiently good overview to realize any such bundle of data structures and methods that is general enough to accommodate all the current and future use cases. Plus: A lot of work has already been put into wrapping Singular rings and modules for Oscar in the first place. Do we really need/want another conversion layer? Or is there, in the end, no other way than to manually get ones hands dirty with the conversions case by case?
> 
> Either way: I am optimistic that localizations can do very well without any direct interface to Singular. At least for the moment.

Sure it is clear that you reduce to the polynomial ring case, so the infrastructure there can be used (your code has new assure functions, but I guess that is not used any more?) 

Generalizing Bi-structures is not really the point. We have there what we want with BiPolyArrays and ModuleGens.

The point is exactly what Claus says. For computations over certain rings, we first translate the problem on the Oscar side to some equivalent mathematical problem, which then gets sent to whatever Gröbner basis machinery. The first step the user should think about, whereas the second one has to happen (and does happen) automatically.

That has to be accommodated not only for the generic approach but for a variety of approaches and also modules over quotient rings (I agree with Claus that in many settings an interface which is based on kernel and coordinates should not be efficient, I would like to see some performance experiments comparing the generic code with standard techniques).

**Comment by fieker** [comment]
Just to put a break on - and because I am lost:


   What actual concrete problem are the two of you trying to solve?

We will never have a fast, efficient, perfect implementation for fin.
gen. modules over Noetherian domains - not even an inefficent one.

We have, currently, to my understanding
 - modules over fields via Gauss, even this is split into
   - modules over fixed size rings (finite field)
   - modules in char 0
 - modules over euclidean rings via Hermite/ Howell
 - modules using Groebner and syz using SubQuo (to avoid Groebner and
   syz)
 - modules over Dedekind using pseudo elements
 - modules over some (semi-)numerical rings (p-adic, ...)
   using clever pivoting and different Gauss
 - modules over localisations that use the modules over Groebner in a 
   clever(?) way

Once we have modules on that level (incl. hom, tensor, sub, quo) and
their homomorphisms, we can build generic mathematics on top, possibly
even efficiently - but I am not too certain.

There are reasons we actually have all those different, independent, but
similar implementations: overwise it is proof-of-concept at most (sure:
HNF and rref (Gauss) could be combined, char 0 and char > 0 fields can
be combined, Gauss is "just" a special version of Groebner, ...)
But fact is, the efficient code uses the underlying machienery as well
as possible. For rref over fin. field: always. For Groebner in char 0:
as rarely as possible. All is better than doing it, hence SubQuo.

So what is the goal?

**Comment by jankoboehm** [comment]
The problem we aim to solve is solely related to modules related in some way to polynomial rings. Nothing to do with the other cases you mention.

It is also solely about the basic layer, the higher level constructions work then as available automatically.

There is now a generic implementation in the case of localizations. There are two questions: Analyze how that compares performance-wise with usual techniques relying on local orderings. And, provide these techniques as a (most likely more performant) alternative to be called in cases they are available. To do so we have to provide some way to implement these techniques in a reasonable way in the existing framework.

**Comment by HechtiDerLachs** [comment]
> Sure it is clear that you reduce to the polynomial ring case, so the infrastructure there can be used (your code has new assure functions, but I guess that is not used any more?)

That's right. The file `src/Modules/LocalizedModules.jl` is not loaded anymore. I just didn't delete it from the git. 

> What actual concrete problem are the two of you trying to solve?

Good question. Janko gave a partial answer already: 

> There is now a generic implementation in the case of localizations. There are two questions: Analyze how that compares performance-wise with usual techniques relying on local orderings. And, provide these techniques as a (most likely more performant) alternative to be called in cases they are available. To do so we have to provide some way to implement these techniques in a reasonable way in the existing framework.

But I don't really see how that requires a new interface or data structure. For instance, if I wanted to use local orderings, because I just happened to localize in a k-point, then I have an associated underlying (Oscar-) polynomial ring and module which is available via `base_ring` and `base_ring_module`. Maybe, one wishes to apply a shift to move the local point to the origin, but that should simply be a change of rings. Either way, the latter modules already have their `singular_assure` methods which I can (hopefully) also invoke with a local ordering. 

The next case we might be considering is the infamous Graal for localizations at complements of prime ideals over `k[y_1, ..., y_m]`. This set of algorithms needs a polynomial ring `K[x_1, ..., x_n]` over a function field `K = k(t_1, ..., t_d)`. So this is not directly available via some `base_ring`, but has to be assembled manually. Also the translation of the original module problem to another module problem over `K[x]` has to be carried out by hand. But then, I expect to really have a module problem over `K[x]` where `K` is a ring that `Singular` can digest and where, again, the standard `singular_assure` methods would also work. Or am I missing something?

Other than that, I can not come up with another example -- probably due to a lack of overview. 

Either way: I am not objecting the introduction of any such new structure in general. All I want to say is 1) Personally, I do not think I need it for anything I plan to do for the schemes, and 2) I still do not get 100% what this is really about, so I am most likely the wrong person to ask for implementing any such thing.

**Comment by jankoboehm** [comment]
OK one more last try. No one is proposing a new data structure. Rather, in addition to our abstract interface to the subquotient framework (membership, lift, kernel), we introduce a second lower level more powerful interface, which the user can choose to implement to make the subquotient framework working. Like the existing abstract interface (which you are using successfully), this is for general use and not just about localizations. What is it about?

To pick up our standard example, when dealing with the localization at a point (or the next case, which is btw not the Graal, but the localization at a hyperplane) there are two steps involved.

Problem A: The problem you want to solve
Problem B: The problem which standard Groebner machines can handle
Problem C: The computation Singular does

To solve our problem we implement A->B->C.

You propose to implement the composition A->C while we want to enforce the user with the interface to clearly define and implement a function A->B (which is Singular independent) with B->C in most cases being automatic. This in particular allows to use alternative Groebner machines.

When you say, you make an OSCAR ring with a local ordering you do nothing else than to realize A->B, so I honestly do not see a disagreement that this approach is how it should be. Perhaps on whether B should be cached, but whether that is necessary and/or useful depends on the problem.

In any case, you do not have to use this interface and can stick with the abstract one, so you don't have to care (though it would be nice if you once available try if it can benefit the performance of your code. And I would like to have a performance comparison between the generic localization code and standard techniques, since this is crucial for our schemes project. That comparison can be also done of course just in an ad hoc way).

**Comment by HechtiDerLachs** [comment]
Alright, thanks for the clarification. 

In this language, what I have implemented with the `Posur.jl` code was more like A->B (but more general and not only in the particular case of polynomial rings). Nowhere in the code I get even close to Singular, i.e. it is really **not** about A->C.

It was my understanding that B->C was handled by `ModuleGens` and their `singular_assure`, at least for the standard case of polynomial rings. I somehow expected that this could then be widened to also digest other base rings that Singular understands. 

If you think that the latter could/should be turned into a documented interface, I am all with you! 

Sure, we should compare the performance of either approaches. But this has also already been done by Barakat et. al. (see Sebastian Posur's paper), so we might ask them for their experiences, first.

**Comment by HechtiDerLachs** [comment]
@AlexD97 : I tried to extend the homomorphisms from SubQuos to also accept change of rings. For some basic examples here it seems to work. But could you also have a look whether you like the implementation and whether it breaks anything?

Btw: Which test-code should I run to check compatibility?

**Comment by HechtiDerLachs** [comment]
Btw: My testing code so far was the following
```julia
R, (x,y) = QQ["x", "y"]
U = MPolyPowersOfElement(R, [x+y, y-x])
S = Localization(U)

Fb = FreeMod(R, 2)
Ab = R[x y]
Mb, pb = quo(Fb, Ab)

F = FreeMod(S, 2)
A = S[x//(x+y)^2 y//(x+y)^2]
M, p = quo(F, (x-y)*A)

f = SubQuoHom(Mb, M, [M[1], M[2]], S)
iswelldefined(f)
iswelldefined(f) # 2 times to check caching
```
Tests in `/test/Modules/UngradedModules.jl` seem to be fine.

**Comment by AlexD97** [comment]
Changes look fine to me. What exactly do you mean with compatibility in 
> Which test-code should I run to check compatibility?

**Comment by HechtiDerLachs** [comment]
Alright, thanks!

> What exactly do you mean with compatibility

The question was: Which tests should I run to be half way sure that my changes don't break anything of your code?

But I guess that `/test/Modules/UngradedModules.jl` does the job, right?

Two other questions: 

1) In order to define a FreeMod over any ring, I need to implement the function `default_ordering(F::FreeMod)`. Shouldn't that be changed for the sake of genericity?

2) This code is now based on a fairly remote branch for the development of the modules. How mergable is that with the original Oscar? Can I rebase my files on some branch of yours that has just the necessary changes to make my code run and can be merged to the official Oscar more easily? I mean: This is a Draft-PR, but I would like to turn it into a serious PR.

**Comment by AlexD97** [comment]
Yes, `/test/Modules/UngradedModules.jl` does the job.

1. Are you sure that you need `default_ordering` in order to define a `FreeMod`?
2. I think my `GroebnerBasis` branch should be mergable but there are a lot of merge-commits. So you think it is better to do a rebase?

**Comment by HechtiDerLachs** [comment]
1) No, the other way around! I am sure, I do **not** need to define a default ordering. But up to now, your code somehow forces me to implement that, nevertheless.

Edit: Sorry, I think I understood wrong. And I was wrong: It was about SubQuos, really. Try running this:
```julia
R, (x,y) = QQ["x","y"]
U = MPolyPowersOfElement(x)
S = Localization(U)
F = FreeMod(S, 2)
A = S[x y]
B = S[x^2 x*y]
M = SubQuo(F, A, B)
```

2) I don't know. We probably should ask @fingolfin , @thofma ? I'm just wondering what is the most direct way to have my code in the official Oscar. If you plan to have your changes merged within next week, then no problem. Otherwise, we need to figure out, how to do it.

**Comment by fingolfin** [comment]
Note that CI tests are not running for this PR as it has conflicts with `master`; a merge or rebase is needed to resolve those before CI runs again.

It seems you asked about this in your last comment. Sorry, this got lost in the flood of GitHub notification emails. If you have such questions, next time better ask on Slack or via email or on gather?

There are 71 commits in here so I am not sure a rebase is feasible. In general I would really advise against such big and long living branches, but oh well, here we are... Next best option might be to just squash this manually into one big commit. This can be done locally on your computer; or we do it when merging (instead of "rebase" or "merge", GitHub allow us to "squash", which produces a single big commit, and that's what I'd use for a PR like this anyway).

**Comment by fingolfin** [comment]
This has been superseded by PR #1295

**Pull request closed without merging**